tiflash: Add desc of new TiFlash storage settings and guidelines for upgrading in v4.0.9 (#4290)
pingcap/docs: tiflash/tiflash-configuration.md

listen_host = "0.0.0.0"
## The TiFlash TCP service port.
tcp_port = 9000
## The TiFlash HTTP service port.
http_port = 8123
## The cache size limit of the metadata of a data block. Generally, you do not need to change this value.
mark_cache_size = 5368709120
## The cache size limit of the min-max index of a data block. Generally, you do not need to change this value.
minmax_index_cache_size = 5368709120

## The storage path of TiFlash data. If there are multiple directories, separate each directory with a comma.
## `path` and `path_realtime_mode` are deprecated since v4.0.9. Use the configurations
## in the `[storage]` section to get better performance in the multi-disk deployment scenarios
# path = "/tidb-data/tiflash-9000"
## or
# path = "/ssd0/tidb-data/tiflash,/ssd1/tidb-data/tiflash,/ssd2/tidb-data/tiflash"
## The default value is `false`. If you set it to `true` and multiple directories 
## are set in the path, the latest data is stored in the first directory and older
## data is stored in the rest directories.
# path_realtime_mode = false 

## The path in which the TiFlash temporary files are stored. By default it is the first directory in `path`

> ## or `storage.latest.dir` appended with "/tmp"

Reviewer: ```suggestion
## or in `storage.latest.dir` appended with "/tmp".
```

# tmp_path = "/tidb-data/tiflash-9000/tmp"

## Storage paths settings since v4.0.9
[storage]
    [storage.main]
    ## The directories to store the main data. More than 90% data is stored in

>     ## these directories.

Author: It means that `dir` is a string list with only one directory.

>     ## these directories.

Reviewer: So should we say "More than 90% of the total data is stored in this directory"?

>     ## these directories.

Reviewer: ```suggestion
    ## the directory list.
```

>     ## The directory(ies) to store the main data. More than 90% of the total data is stored in

Reviewer: ```suggestion
    ## The directory(ies) to store the main data. More than 90% of the total data is stored in
```

>     ## The directory(ies) to store the main data. More than 90% of the total data is stored in

Reviewer: ```suggestion
    ## The list of directories to store the main data. More than 90% of the total data is stored in
```

>     ## these directories.

Reviewer: ```suggestion
    ## `dir`.
```

    dir = [ "/tidb-data/tiflash-9000" ] 
    ## or
    # dir = [ "/ssd0/tidb-data/tiflash", "/ssd1/tidb-data/tiflash" ]

    ## The maximum storage capacity of each directory in `storage.main.dir`.
    ## If it is not set, or is set to multiple 0, the actual disk (the disk where the directory is located) capacity is used.

>     ## Note that big numbers such as "10GB" are not supported yet.

Author: This is for those users who see TiKV configuration. They may think that TiFlash supports this definition of a human-readable number too. But actually, TiFlash does not.
```suggestion
    ## Note that human-readable numbers such as "10GB" are not supported yet.
```

>     ## Set the specified number in bytes.

Reviewer: ```suggestion
    ## Numbers are specified in bytes.
```

    ## The size of the `capacity` list should be the same with the `dir` size.
    ## For example:
    # capacity = [ 10737418240, 10737418240 ]

    [storage.latest]

>     ## The directories to store the latest data. About 10% of the total data is stored in

Reviewer: ```suggestion
    ## The directory(ies) to store the latest data. About 10% of the total data is stored in
```

>     ## these directories. The directories listed here require higher IOPS

Reviewer: ```suggestion
    ## the directory list. The directories (or directory) listed here require higher IOPS
```

mark_cache_size = 5368709120
## The cache size limit of the min-max index of a data block. Generally, you do not need to change this value.
minmax_index_cache_size = 5368709120
## The cache size limit of the DeltaIndex. The default value is 0, which means no limit.
delta_index_cache_size = 0

## The storage path of TiFlash data. If there are multiple directories, separate each directory with a comma.
## `path` and `path_realtime_mode` are deprecated since v4.0.9. Use the configurations
## in the `[storage]` section to get better performance in the multi-disk deployment scenarios
# path = "/tidb-data/tiflash-9000"
## or
# path = "/ssd0/tidb-data/tiflash,/ssd1/tidb-data/tiflash,/ssd2/tidb-data/tiflash"
## The default value is `false`. If you set it to `true` and multiple directories
## are set in the path, the latest data is stored in the first directory and older
## data is stored in the rest directories.
# path_realtime_mode = false

## The path in which the TiFlash temporary files are stored. By default it is the first directory in `path`
## or in `storage.latest.dir` appended with "/tmp".
# tmp_path = "/tidb-data/tiflash-9000/tmp"

> ## Storage paths settings effective since v4.0.9

Author: ```suggestion
## Storage paths settings take effect starting from v4.0.9
```

    ## properly increase the latency to reduce IOPS usage. The default value is "4ms".
    ## If you set it to 0ms, the optimization is disabled.
    store-batch-retry-recv-timeout = "4ms"
```

In addition to the items above, other parameters are the same with those of TiKV. Note that the configuration items in `tiflash.toml [flash.proxy]` will override the overlapping parameters in `tiflash-learner.toml`; The `label` whose key is `engine` is reserved and cannot be configured manually.

### Multi-disk deployment

TiFlash supports multi-disk deployment. If there are multiple disks in your TiFlash node, you can make full use of those disks by following configurations. You can check the template for TiUP deployment here: [The complex template for the TiFlash topology](https://github.com/pingcap/docs/blob/master/config-templates/complex-tiflash.yaml). 

#### Multi-disk deployment with version < v4.0.9

For TiDB clusters with version lower than v4.0.9, TiFlash only supports storing the main data of the storage engine on multiple disks. You can setup a TiFlash node on multiple disks by specifying the `path` (`data_dir` in TiUP) and `path_realtime_mode` configuration.

> If there are multiple directories to store data, join those directories with a comma and set it to `path`. For example, `/nvme_ssd_a/data/tiflash,/sata_ssd_b/data/tiflash,/sata_ssd_c/data/tiflash`. If there are multiple disks on your TiFlash node, it is recommended to put disks with the best I/O metrics at the front of `path` to maximize the performance of the node.

Reviewer: The deleted version looks nice~
```suggestion
If there are multiple data storage directories in `path`, separate each with a comma. For example, `/ssd_a/data/tiflash,/hdd_b/data/tiflash,/hdd_c/data/tiflash`. If there are multiple disks in your environment, it is recommended that each directory corresponds to one disk and you put disks with the best performance at the front to maximize the performance of all disks.
```

> If there are multiple disks with similar I/O metrics on your TiFlash node, you can leave the `path_realtime_mode` parameter to be the default value (or you can specifically set it to `false`). It means that data are evenly distributed on all storage directories. However the latest data is written to the first directory, the corresponding disk is still busier than other disks.

Reviewer: ```suggestion
If there are multiple disks with similar I/O metrics on your TiFlash node, you can leave the `path_realtime_mode` parameter to the default value (or you can explicitly set it to `false`). It means that data will be evenly distributed among all storage directories. However, the latest data is written only to the first directory, so the corresponding disk is busier than other disks.
```

> If there are multiple disks with different I/O metrics on your TiFlash node, it is recommended to set `path_realtime_mode` to `true` and put disks with the best I/O metrics at the front of `path`. It means that the first directory only stores the latest data, and the older data are evenly distributed on the other directories. Notice that in this case, the capacity of the first directory should be planned as 10% of the total capacity of all directories.

Reviewer: ```suggestion
If there are multiple disks with different I/O metrics on your TiFlash node, it is recommended to set `path_realtime_mode` to `true` and put disks with the best I/O metrics at the front of `path`. It means that the first directory only stores the latest data, and the older data are evenly distributed among the other directories. Note that in this case, the capacity of the first directory should be planned as 10% of the total capacity of all directories.
```

> #### Multi-disk deployment with version >= v4.0.9

Reviewer: ```suggestion
#### Multi-disk deployment with TiDB v4.0.9 or later
```

> For TiDB clusters with version v4.0.9 and higher, TiFlash supports storing the main data and the latest data of the storage engine on multiple disks. If you want to deploy a TiFlash node on multiple disks, it is recommended to specify your storage directories by the `[storage]` section to make full use of your node. However, the configurations before v4.0.9 (`path` and `path_realtime_mode`) are still supported.

Reviewer: ```suggestion
For TiDB clusters with v4.0.9 or later versions, TiFlash supports storing the main data and the latest data of the storage engine on multiple disks. If you want to deploy a TiFlash node on multiple disks, it is recommended to specify your storage directories in the `[storage]` section to make full use of your node. Note that the configurations earlier than v4.0.9 (`path` and `path_realtime_mode`) are still supported.
```

> If there are multiple disks with similar I/O metrics on your TiFlash node, it is recommended to specify corresponds directories in the `storage.main.dir` list and leave `storage.latest.dir` to be empty. TiFlash will distribute I/O pressure and data placement among all directories.

Reviewer: ```suggestion
If there are multiple disks with similar I/O metrics on your TiFlash node, it is recommended to specify corresponding directories in the `storage.main.dir` list and leave `storage.latest.dir` empty. TiFlash will distribute I/O pressure and data among all directories.
```

> If there are multiple disks with different I/O metrics on your TiFlash node, it is recommended to specify directories with higher metrics in the `storage.latest.dir` list, and specify directories with lower metrics in the `storage.main.dir` list. For example, there are one NVME-SSD and two SATA-SSDs, you can set `storage.latest.dir` to `["/nvme_ssd_a/data/tiflash"]` and `storage.main.dir` to `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`. TiFlash will distribute I/O pressure and data placement among these two directories list separately. Notice that in this case, the capacity of `storage.latest.dir` should be planned as 10% of the total planned capacity.

Reviewer: ```suggestion
If there are multiple disks with different I/O metrics on your TiFlash node, it is recommended to specify directories with higher metrics in the `storage.latest.dir` list, and specify directories with lower metrics in the `storage.main.dir` list. For example, for one NVMe-SSD and two SATA-SSDs, you can set `storage.latest.dir` to `["/nvme_ssd_a/data/tiflash"]` and `storage.main.dir` to `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`. TiFlash will distribute I/O pressure and data among these two directories list respectively. Note that in this case, the capacity of `storage.latest.dir` should be planned as 10% of the total planned capacity.
```

> > **Notes:**

Reviewer: ```suggestion
> **Warning:**
```

>

> > * The [storage] configuration is supported in TiUP since v1.2.5. If your TiDB cluster version is not less than v4.0.9, please make sure that your TiUP version is not less than v1.2.5. Or the data directories defined in [storage] won't be managed by TiUP.

Reviewer: ```suggestion
> * The `[storage]` configuration is supported in TiUP since v1.2.5. If your TiDB cluster v4.0.9 or later, make sure that your TiUP version is v1.2.5 or later. Otherwise, the data directories defined in `[storage]` will not be managed by TiUP.
```

[server]
    engine-addr = The external access address of the TiFlash coprocessor service.
[raftstore]
    ## Specifies the number of threads that handle snapshots.
    ## The default number is 2. 
    ## If you set it to 0, the multi-thread optimization is disabled.
    snap-handle-pool-size = 2 

    ## Specifies the shortest interval at which Raft store persists WAL.
    ## You can properly increase the latency to reduce IOPS usage.
    ## The default value is "4ms".
    ## If you set it to 0ms, the optimization is disabled.
    store-batch-retry-recv-timeout = "4ms"
```

In addition to the items above, other parameters are the same with those of TiKV. Note that the configuration items in `tiflash.toml [flash.proxy]` will override the overlapping parameters in `tiflash-learner.toml`; The `label` whose key is `engine` is reserved and cannot be configured manually.

### Multi-disk deployment

> TiFlash supports multi-disk deployment. If there are multiple disks in your TiFlash node, you can make full use of those disks by configuring the parameters described in the following sections. For the configuration template used for TiUP, see [The complex template for the TiFlash topology](https://github.com/pingcap/docs/blob/master/config-templates/complex-tiflash.yaml). 

Reviewer: ```suggestion
TiFlash supports multi-disk deployment. If there are multiple disks in your TiFlash node, you can make full use of those disks by configuring the parameters described in the following sections. For TiFlash's configuration template to be used for TiUP, see [The complex template for the TiFlash topology](https://github.com/pingcap/docs/blob/master/config-templates/complex-tiflash.yaml). 
```

#### Multi-disk deployment with TiDB version earlier than v4.0.9

For TiDB clusters earlier than v4.0.9, TiFlash only supports storing the main data of the storage engine on multiple disks. You can set up a TiFlash node on multiple disks by specifying the `path` (`data_dir` in TiUP) and `path_realtime_mode` configuration.

> If there are multiple data storage directories in `path`, separate each with a comma. For example, `/ssd_a/data/tiflash,/hdd_b/data/tiflash,/hdd_c/data/tiflash`. If there are multiple disks in your environment, it is recommended that each directory corresponds to one disk and you put disks with the best performance at the front to maximize the performance of all disks.

Author: ```suggestion
If there are multiple data storage directories in `path`, separate each with a comma. For example, `/nvme_ssd_a/data/tiflash,/sata_ssd_b/data/tiflash,/sata_ssd_c/data/tiflash`. If there are multiple disks in your environment, it is recommended that each directory corresponds to one disk and you put disks with the best performance at the front to maximize the performance of all disks.
```

If there are multiple disks with similar I/O metrics on your TiFlash node, it is recommended to specify corresponding directories in the `storage.main.dir` list and leave `storage.latest.dir` empty. TiFlash will distribute I/O pressure and data among all directories.

If there are multiple disks with different I/O metrics on your TiFlash node, it is recommended to specify directories with higher metrics in the `storage.latest.dir` list, and specify directories with lower metrics in the `storage.main.dir` list. For example, for one NVMe-SSD and two SATA-SSDs, you can set `storage.latest.dir` to `["/nvme_ssd_a/data/tiflash"]` and `storage.main.dir` to `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`. TiFlash will distribute I/O pressure and data among these two directories list respectively. Note that in this case, the capacity of `storage.latest.dir` should be planned as 10% of the total planned capacity.

> **Warning:**
>

> > * The `[storage]` configuration is supported in TiUP since v1.2.5. If your TiDB cluster v4.0.9 or later, make sure that your TiUP version is v1.2.5 or later. Otherwise, the data directories defined in `[storage]` will not be managed by TiUP.

Author: ```suggestion
> * The `[storage]` configuration is supported in TiUP since v1.2.5. If your TiDB cluster version is v4.0.9 or later, make sure that your TiUP version is v1.2.5 or later. Otherwise, the data directories defined in `[storage]` will not be managed by TiUP.
```

> > * After turning to use the [storage] configurations, downgrading your cluster version to less than v4.0.9 may make some TiFlash data lost.

Reviewer: ```suggestion
> * After using the [storage] configurations, downgrading your cluster to a version earlier than v4.0.9 might cause **data loss** on TiFlash..
```
